packages/web3-eth-contract/src/index.js

```diff
@@ -581,9 +581,13 @@ Contract.prototype.deploy = function(options, callback){
     options = this._getOrSetDefaultOptions(options);
 
 
-    // return error, if no "data" is specified
+    // throw error, if no "data" is specified
     if(!options.data) {
-        return utils._fireError(new Error('No "data" specified in neither the given options, nor the default options.'), null, null, callback);
+        if (typeof callback === 'function'){
+            callback(errors.ContractMissingDeployDataError());
+        } else {
+            throw errors.ContractMissingDeployDataError();
+        }
```

Comment:
This block is the only real logic change. Fixes bug where `utils._fireError` is usually swallowing this error.

`contract.deploy` doesn't have an event emitter or a promise associated with it (and AFAIK, isn't really invoked with a callback) 

Have rewritten it into a thrown error but kept backwards compat for the callback in case anyone actually doing that.